pkg/kubecost/allocation.go

```diff
@@ -1539,7 +1542,56 @@ func (as *AllocationSet) AggregateBy(aggregateBy []string, options *AllocationAg
 					if alloc.SharedCostBreakdown == nil {
 						alloc.SharedCostBreakdown = map[string]SharedCostBreakdown{}
 					}
-					sharedCostName := sharedAlloc.generateKey(aggregateBy, options.LabelConfig)
+
+					var sharedAggregateBy []string
+					var sharedLabels [][]string
+					for labelKey, labelValues := range options.SharedLabels {
+						sharedAgg := fmt.Sprintf("label:%s", labelKey)
+						if !slices.Contains(sharedAggregateBy, sharedAgg) {
+							sharedAggregateBy = append(sharedAggregateBy, sharedAgg)
+						}
+						sharedLabels = append(sharedLabels, labelValues)
+					}
+					if len(options.SharedNamespaces) > 0 {
+						sharedAggregateBy = append(sharedAggregateBy, "namespace")
+					}
+					sharedCostName := sharedAlloc.generateKey(sharedAggregateBy, options.LabelConfig)
+
+					// get each value in the generated key, then reset the name
+					sharedCostNameValues := strings.Split(sharedCostName, "/")
+					sharedCostName = ""
+
+					// if we don't have as many values as aggregateBys, something went wrong in generateKey
+					if len(sharedCostNameValues) != len(sharedAggregateBy) {
+						log.Warnf("Unable to determine share cost group for allocation \"%s\"", sharedAlloc.Name)
+						sharedCostName = "unknown"
+					} else {
+						// try to match to the first label
+						for i, sharedLabelValues := range sharedLabels {
+							allocLabel := sharedCostNameValues[i]
+							if slices.Contains(sharedLabelValues, allocLabel) {
+								sharedCostName = allocLabel
+								break
+							}
+						}
+
+						// if we didn't match to a label, try to match to a namespace
```

Comment:
I think we'll do this even if we do match a label. Probably very minor, but might be something to try to fix nevertheless.

Have you considered making this whole chunk of code a function so that you can early return? E.g.
```go
func determineSharingName(sharedAlloc, options) (string, error)
```

Two advantages to that:
1. Easier to unit test just this little portion of logic
2. Can early return, instead of trying to break in funny ways